rust_snuba/src/mutations/clickhouse.rs

```diff
@@ -0,0 +1,158 @@
+        table: &str,
+        database: &str,
+        clickhouse_user: &str,
+        clickhouse_password: &str,
+        batch_write_timeout: Option<Duration>,
+    ) -> Self {
+        let mut headers = HeaderMap::with_capacity(5);
+        headers.insert(CONNECTION, HeaderValue::from_static("keep-alive"));
+        headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip,deflate"));
+        headers.insert(
+            "X-Clickhouse-User",
+            HeaderValue::from_str(clickhouse_user).unwrap(),
+        );
+        headers.insert(
+            "X-ClickHouse-Key",
+            HeaderValue::from_str(clickhouse_password).unwrap(),
+        );
+        headers.insert(
+            "X-ClickHouse-Database",
+            HeaderValue::from_str(database).unwrap(),
+        );
+
+        let url = format!("http://{hostname}:{http_port}");
+
+        let mut client_builder = ClientBuilder::new().default_headers(headers);
+
+        if let Some(timeout) = batch_write_timeout {
+            client_builder = client_builder.timeout(timeout);
+        }
+
+        Self {
+            url,
+            table: table.to_owned(),
+            client: client_builder.build().unwrap(),
+        }
+    }
+
+    async fn process_message(&self, message: &Message<MutationBatch>) -> anyhow::Result<()> {
+        let body = format_query(&self.table, message.payload());
+
+        self.client
+            .post(&self.url)
+            .body(body)
+            .send()
+            .await?
+            .error_for_status()?;
+
+        Ok(())
+    }
+}
+
+impl TaskRunner<MutationBatch, (), anyhow::Error> for ClickhouseWriter {
+    fn get_task(&self, message: Message<MutationBatch>) -> RunTaskFunc<(), anyhow::Error> {
+        let slf = self.clone();
+
+        Box::pin(async move {
+            slf.process_message(&message)
+                .await
+                .map_err(RunTaskError::Other)?;
+            message.try_map(|_| Ok(()))
+        })
+    }
+}
+
+fn format_query(table: &str, batch: &MutationBatch) -> Vec<u8> {
+    let mut attr_columns = String::new();
+    // attr_combined_columns is intentionally ordered the same as the clickhouse schema.
+    // INSERT INTO .. SELECT FROM .. matches up columns by position only, and ignores names.
+    // subqueries don't have this property.
+    let mut attr_combined_columns = String::new();
+    for i in 0..ATTRS_SHARD_FACTOR {
+        attr_columns.push_str(&format!(",attr_str_{i} Map(String, String)"));
+
+        attr_combined_columns.push_str(&format!(
+            ",mapUpdate(old_data.attr_str_{i}, new_data.attr_str_{i}) AS attr_str_{i}"
+        ));
+    }
+
+    for i in 0..ATTRS_SHARD_FACTOR {
+        attr_columns.push_str(&format!(",attr_num_{i} Map(String, Float64)"));
+        attr_combined_columns.push_str(&format!(
+            ",mapUpdate(old_data.attr_num_{i}, new_data.attr_num_{i}) AS attr_num_{i}"
+        ));
+    }
+
+    let mut body = format!(
+        "
+        INSERT INTO {table}
```

Comment:
@untitaker 2 thoughts on this query:

1. We can rewrite this to avoid JOIN by using the WITH context. We would need to look into the performance differences. 

2. What if we tried a UNION to get both the cancellation and the update in this one query? (SELECT ... UNION SELECT ...)

I'm also wondering if there's a way to do something like INSERT INTO ... VALUES (SELECT ...) (SELECT ...) so we can have 2 insertions back to back? 

Just some thoughts -- I can try these out on some local data

Reply:
I think WITH conflicts directly with input() in my tests. I was also trying to use WITH to make the cancellation rows happen (using both WITH and UNION ALL). I'm not sure how it would help to avoid the join?

If you have a specific query let's discuss 1:1 or just push it, if it works already.

> so we can have 2 insertions back to back?

then we should just merge those MutationBatches at JSON level, I don't think the query needs to adapt to accomodate larger batches.